File: Data Structures & Algorithms/combination-target-sum/submission-3.cpp
class Solution {
    vector<vector<int>> res;
    void dfs(int i, vector<int>& cur, int total, vector<int>& nums, int target) {
        if (total == target) {
            res.push_back(cur);
            return;
        }

        // Loop through all next choices
        for (int j = i; j < nums.size(); j++) {
            if (total + nums[j] > target)
                return;
            cur.push_back(nums[j]);
            dfs(j, cur, total + nums[j], nums, target);
            cur.pop_back();
        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> cur;
        sort(nums.begin(), nums.end());
        dfs(0, cur, 0, nums, target);
        return res;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-3.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        count = {}

        for x in nums:
            count[x] = count.get(x, 0) + 1

        for i, a in enumerate(nums):
            count[a] -= 1
            if i and nums[i] == nums[i-1]:
                continue
            
            for j in range(i+1, len(nums)):
                count[nums[j]] -= 1
                if j-1>i and nums[j] == nums[j-1]:
                    continue
                target = -a - nums[j]
                if count.get(target, 0) != 0:
                    res.append([a, nums[j], target])
            
            for j in range(i+1, len(nums)):
                count[nums[j]] += 1

        return res